### ISSUE TITLE
+= 1 counter missing when build with --prod

### ISSUE THREAD

**Issue by lonelywm** [angular/angular-cli#16144]
The code is ok when i run "ng serve", but when I publish it  with "ng build --prod", something is wrong
It took me a long time to find where the err come from, and eventually I found it.

origin code: 
PS. A function to find whether the item is in nodes and return the index.
```
function(item,nodes){
    let index = -1;
    for (const node of nodes) {
      index += 1;
      if (item === node) {
        return index;
      }
    }
    return -1;
  }
```

After the compiler's optimazation and zip the corresponding code is:
if always return 0 or -1 because the variable 'e' never plus itself by 1
```
function(n,l){
for(var e=-1,t=0,i=l;t<i.length;t++)
if(n===i[t])return e+1;return-1}
}
```

## 🌍  My Environment
    "@angular-devkit/build-angular": # test on "0.802.2" && "0.83.18"
    "@angular/animations": "^8.2.6",
    "@angular/common": "8.2.2",
    "@angular/compiler": "8.2.2",
    "@angular/core": "8.2.2",
    "@angular/forms": "8.2.2",
    "@angular/platform-browser": "8.2.2",
    "@angular/platform-browser-dynamic": "8.2.2",

**Comment by clydin** [comment]
Can you provide the output of `npm ls terser`?

If interested and assuming `nodes` is an Array, that entire function could be replaced with `nodes.findIndex(....)`.  It is polyfill-ed automatically by the CLI as it is used by Angular itself so it is safe to use for ES5 browsers

**Comment by lonelywm** [comment]
xxxxxxx@0.0.0 D:\_xxxxxx
`-- @angular-devkit/build-angular@0.803.18
  +-- terser@4.3.9
  `-- terser-webpack-plugin@1.4.1
    `-- terser@4.3.9  deduped

**Comment by lonelywm** [comment]
Thank you, I've learned.

**Comment by filipesilva** [comment]
I checked with both `terser@4.3.9` and `4.4.0` (latest) and the output for both was `var func=function(r,n){for(var f=-1,t=0,u=n;t<u.length;t++){if(f+=1,r===u[t])return f}return-1};`

My setup was this:
- `package.json`
```
{
  "name": "random",
  "scripts": {
    "test": "tsc input.ts && terser --compress --mangle -- input.js"
  },
  "dependencies": {
    "terser": "4.4.0",
    "typescript": "3.7.2"
  }
}
```
- `input.ts`
```
var func = function (item, nodes) {
  let index = -1;
  for (const node of nodes) {
    index += 1;
    if (item === node) {
      return index;
    }
  }
  return -1;
}
```

Not sure what's happening. Might be a weird interaction in our separate terser compress/mangle?

**Comment by lonelywm** [comment]
Thanks for your patience, I have replaced the function with old style for-loop "(let i = 0; i < xx.length; i++) ",  everything works well. Maybe this err is caused by complex context.

**Comment by filipesilva** [comment]
Tried reproducing outside of terser itself and was able to.

I added the code below to `app.component.ts`:

```
const myFn = function (item, nodes) {
  let index = -1;
  for (const node of nodes) {
    index += 1;
    if (item === node) {
      return index;
    }
  }
  return -1;
}

const nodes = [1, 2, 3];
console.log(myFn(0, nodes), myFn(1, nodes), myFn(2, nodes), myFn(3, nodes), myFn(4, nodes));
```

Running `ng serve` will show `-1 0 1 2 -1` on the console, but `ng serve --prod` will instead show `-1 0 0 0 -1`. 

So there definitely is something in our pipeline causing this. Will look into it.

Versions in my repro:
```
Angular CLI: 9.0.0-rc.1
Node: 10.16.0
OS: win32 x64
Angular: 9.0.0-rc.1
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.0-rc.1
@angular-devkit/build-angular     0.900.0-rc.1
@angular-devkit/build-optimizer   0.900.0-rc.1
@angular-devkit/build-webpack     0.900.0-rc.1
@angular-devkit/core              9.0.0-rc.1
@angular-devkit/schematics        9.0.0-rc.1
@ngtools/webpack                  9.0.0-rc.1
@schematics/angular               9.0.0-rc.1
@schematics/update                0.900.0-rc.1
rxjs                              6.5.3
typescript                        3.6.4
webpack                           4.41.2
```
```
kamik@RED-X1C6 MINGW64 /d/sandbox/next-app (master)
$ npm ls terser
next-app@0.0.0 D:\sandbox\next-app
`-- @angular-devkit/build-angular@0.900.0-rc.1
  +-- terser@4.3.9 
  +-- terser-webpack-plugin@2.2.1
  | `-- terser@4.3.9  deduped
  `-- webpack@4.41.2
    `-- terser-webpack-plugin@1.4.1
      `-- terser@4.3.9  deduped

kamik@RED-X1C6 MINGW64 /d/sandbox/next-app (master)
$ npm ls @babel/core
next-app@0.0.0 D:\sandbox\next-app
`-- @angular-devkit/build-angular@0.900.0-rc.1
  `-- @babel/core@7.6.4 
```

**Comment by clydin** [comment]
This can be replicated using the above code with terser only and appears to be a variation of this terser issue: https://github.com/terser/terser/issues/475.  The `++` case was fixed but `+= 1` appears to still fail.

**Comment by filipesilva** [comment]
Note: I changed the issue title to better reflect the problem.

**Comment by lonelywm** [comment]
Nice work!

**Comment by filipesilva** [comment]
At this point we're waiting for a Terser fix, and will pick it up automatically when available.

**Comment by clydin** [comment]
As per the referenced terser issue, this has been corrected in terser 4.5+.  This version will be available in 9.0.
